### ISSUE TITLE
Bug: Incorrect dtype if training with the AMP after loss calculation

### ISSUE THREAD

**Issue by guyleaf** [lyuwenyu/RT-DETR#634]
**Star RTDETR**
请先在RTDETR主页点击**star**以支持本项目
Star RTDETR to help more people discover this project. 

## Description
**Related comment**
According to the https://github.com/lyuwenyu/RT-DETR/pull/424#issuecomment-2295530965, you mentioned
> It's used to make sure pure float32 for loss during the `criterion` phase. And I'm not sure the mAP result when `enabled=True`. Can you check it?

But current implementation will not work as expected.

**Describe the bug**
Current implementation will cause `loss_vfl` is calculated in `torch.float16`.

<img width="706" height="338" alt="Image" src="https://github.com/user-attachments/assets/4b60e328-9943-45a9-8aa0-ac734027308e" />

**To Reproduce**
1. Modify the code like below in [`det_engine.py`](https://github.com/lyuwenyu/RT-DETR/blob/f9417e3acfa48bcb649e5ec0bc3de1e8677c8961/rtdetrv2_pytorch/src/solver/det_engine.py#L44).
```python
if scaler is not None:
    with torch.autocast(device_type=str(device), cache_enabled=True):
        outputs = model(samples, targets=targets)
    
    with torch.autocast(device_type=str(device), enabled=False):
        loss_dict = criterion(outputs, targets, **metas)

    print(loss_dict)
    exit()
```
2. Train with `--use-amp` with any config.

## Possible solution
According to [the official guide of using AMP](https://docs.pytorch.org/tutorials/recipes/recipes/amp_recipe.html#adding-torch-autocast), the criterion will autocast to `float32`.

So, I modify the code like below, then it works.
```python
if scaler is not None:
    with torch.autocast(device_type=str(device), cache_enabled=True):
        outputs = model(samples, targets=targets)
        loss_dict = criterion(outputs, targets, **metas)
    
    # with torch.autocast(device_type=str(device), enabled=False):
    #    loss_dict = criterion(outputs, targets, **metas)

    print(loss_dict)
    exit()
```

<img width="571" height="347" alt="Image" src="https://github.com/user-attachments/assets/7e47ee04-80fa-4056-9036-784fd2990fcf" />

Otherwise, we have to cast them manually.

If it is accepted, I can create a Pull Request to fix it.
